File: Tests/TestSchroeder.py
"""
University of Minnesota
Aerospace Engineering and Mechanics - UAV Lab
Copyright 2019 Regents of the University of Minnesota
See: LICENSE.md for complete license details

Author: Chris Regan

Example script for generating Schroeder type Multisine Excitations.
"""

import numpy as np
import matplotlib.pyplot as plt

# Hack to allow loading the Core package
if __name__ == "__main__" and __package__ is None:
    from sys import path, argv
    from os.path import dirname, abspath, join

    path.insert(0, abspath(join(dirname(argv[0]), "..")))
    path.insert(0, abspath(join(dirname(argv[0]), "..", 'Core')))
    
    del path, argv, dirname, abspath, join

from Core import GenExcite
from Core import FreqTrans


# Constants
hz2rps = 2*np.pi
rps2hz = 1/hz2rps

#%% Define the frequency selection and distribution of the frequencies into the signals
numChan = 3
freqRate_hz = 50;
timeDur_s = 10.0
numCycles = 5

freqMinDes_rps = (numCycles/timeDur_s) * hz2rps * np.ones(numChan)
#freqMaxDes_rps = (freqRate_hz/2) * hz2rps *  np.ones(numChan)
freqMaxDes_rps = 10 * hz2rps *  np.ones(numChan)
freqStepDes_rps = (5 / freqRate_hz) * hz2rps
methodSW = 'zip' # "zippered" component distribution

## Generate MultiSine Frequencies
freqElem_rps, sigIndx, time_s = GenExcite.MultiSineComponents(freqMinDes_rps, freqMaxDes_rps, freqRate_hz, numCycles, freqStepDes_rps, methodSW)
timeDur_s = time_s[-1] - time_s[0]

## Generate Schroeder MultiSine Signal
ampElem_nd = np.ones_like(freqElem_rps) ## Approximate relative signal amplitude, create flat
sigList, phaseElem_rad, sigElem = GenExcite.MultiSine(freqElem_rps, ampElem_nd, sigIndx, time_s, costType = 'Schroeder', phaseInit_rad = 0, boundPhase = 1, initZero = 1, normalize = 'peak');


## Results
peakFactor = GenExcite.PeakFactor(sigList)
peakFactorRel = peakFactor / np.sqrt(2)
print(peakFactorRel)

# Signal Power
sigPowerRel = (ampElem_nd / max(ampElem_nd))**2 / len(ampElem_nd)



if True:
    fig, ax = plt.subplots(ncols=1, nrows=numChan, sharex=True)
    for iChan in range(0, numChan):
        ax[iChan].plot(time_s, sigList[iChan])
        ax[iChan].set_ylabel('Amplitude (nd)')
        ax[iChan].grid(True)
    ax[iChan].set_xlabel('Time (s)')

if True:
    fig, ax = plt.subplots(ncols=1, nrows=numChan, sharex=True)
    for iChan in range(0, numChan):
        for iElem in sigIndx[iChan]:
            ax[iChan].plot(time_s, sigElem[iElem])
        ax[iChan].set_ylabel('Amplitude (nd)')
        ax[iChan].grid(True)
    ax[iChan].set_xlabel('Time (s)')


#%% Plot the Excitation Spectrum

## Compute Spectrum of each channel
optFFT = FreqTrans.OptSpect(freqRate = freqRate_hz * hz2rps)
optCZT = FreqTrans.OptSpect(dftType = 'czt', freqRate = freqRate_hz * hz2rps)

freq_fft = []
P_dB_fft = []
freq_czt = []
P_dB_czt = []

for iChan, sig in enumerate(sigList):
    freq_rps_fft, _, P_fft  = FreqTrans.Spectrum(sig, optFFT)
    freq_fft.append(freq_rps_fft * rps2hz)
    P_dB_fft.append(20*np.log10(P_fft))
    
    optCZT.freq = freqElem_rps[sigIndx[iChan]]
    freq_rps_czt, _, P_czt  = FreqTrans.Spectrum(sig, optCZT)
    freq_czt.append(freq_rps_czt * rps2hz)
    P_dB_czt.append(20*np.log10(P_czt))
    
nChan = len(P_dB_fft)

plt.figure()
for iChan in range(0, nChan):
    plt.subplot(nChan, 1, iChan+1)
    plt.plot(freq_fft[iChan].T, P_dB_fft[iChan].T, '-k', label='FFT Pxx')
    plt.plot(freq_czt[iChan].T, P_dB_czt[iChan].T, '.r-', label='CZT Pxx')
    plt.grid()
    plt.ylabel('Spectrum (dB)');

plt.xlabel('frequency (Hz)');
plt.legend()
plt.show()


#%% Create the output for JSON config

timeFinal_s = time_s[-1]
timeStart_s = time_s[0]

jsonMulti = {}
for iChan in range(0, numChan):
    iElem = sigIndx[iChan]
        
    dictChan = {}
    dictChan['Type'] = 'MultiSine'
    dictChan['Duration'] = timeDur_s
    dictChan['Frequency'] = list(freqElem_rps[iElem])
    dictChan['Phase'] = list(phaseElem_rad[iElem])
    dictChan['Amplitude'] = list(ampElem_nd[iElem])

    nameChan = 'OMS_' + str(iChan+1)
    
    jsonMulti[nameChan] = dictChan

import json
print(json.dumps(jsonMulti, separators=(', ', ': ')))

File: Tests/TestSchroeder.m
% Test Shroeder MultiSine Generation
%
% Notes:
%

% University of Minnesota
% Aerospace Engineering and Mechanics - UAV Lab
% Copyright (c) 2019 Regents of the University of Minnesota
% See: LICENSE.md for complete license details
% Author: Chris Regan
%

%% TODO: Add comments and convert to validation form
clear all;

% Constants
hz2rps = 2*pi;
rps2hz = 1/hz2rps;


%% Define the frequency selection and distribution of the frequencies into the signals
structMultiSine.numChan = 3;
structMultiSine.timeRate_s = 1/50;
structMultiSine.timeDur_s = 10.0;
structMultiSine.numCycles = 2;

freqMinDes_hz = structMultiSine.numCycles / structMultiSine.timeDur_s;
freqMaxDes_hz = 6.0;

structMultiSine.freqRange_rps = repmat([freqMinDes_hz, freqMaxDes_hz], [structMultiSine.numChan, 1]) * hz2rps;
structMultiSine.freqStepDes_rps = (1 / 10) * hz2rps;
methodSW = 'zip'; % "zippered" component distribution

structMultiSine = MultiSineComponents(structMultiSine, methodSW);


%% Amplitude
structMultiSine.ampChan_nd = {};
for iChan = 1:structMultiSine.numChan
    structMultiSine.ampChan_nd{iChan} = ones(size(structMultiSine.freqChan_rps{iChan})); % Flat-spectrum, each signal has unity amplitude
end


%% Generate Schoeder MultiSine Signal
phaseComp1_rad = [];
boundSW = [];
normalSW = true;
forceZeroSW = true;

[structMultiSine] = MultiSineSchroeder(structMultiSine, phaseComp1_rad, boundSW, normalSW, forceZeroSW);


%% Modify the Excitation Signal
numRepeat = [];
timeLead_s = 1*structMultiSine.timeRate_s;
timeTrail_s = 1*structMultiSine.timeRate_s;

[structMultiSine] = ModifyExcitation(structMultiSine, numRepeat, timeLead_s, timeTrail_s);


%% Results
signals = structMultiSine.signals;
peakFactorRel = PeakFactor(signals)/sqrt(2)

figure;
for iChan = 1:structMultiSine.numChan
    subplot(structMultiSine.numChan, 1, iChan);
    plot(structMultiSine.time_s, structMultiSine.signals(iChan, :)); grid on;
end

% PSD
structMultiSine.freqRate_rps = 1/structMultiSine.timeRate_s * hz2rps;
winType = 'rect';
smoothFactor = 1;

% [xxP, xDft, freq_rps] = SpectEst(signals, freqRate_rps, [], winType, smoothFactor, 'FFT');
% SpectPlot(freq_rps, xxP, 'rad/sec', [], []);

% Chrip-Z
xxP_CZ = {};
xDft_CZ = {};
freq_rps_CZ = {};
for iChan = 1:structMultiSine.numChan
    [xxP_CZ{iChan}, xDft_CZ{iChan}, freq_rps_CZ{iChan}] = SpectEst(structMultiSine.signals, structMultiSine.freqRate_rps, structMultiSine.freqChan_rps{iChan}, winType, smoothFactor, 'ChirpZ');
end

% SpectPlot(freq_rps_CZ, xxP_CZ, 'rad/sec', [], []);
figure;
for iChan = 1:structMultiSine.numChan
    subplot(structMultiSine.numChan, 1, iChan);
    semilogx(freq_rps_CZ{iChan}, Mag2DB( xxP_CZ{iChan} ), '*-'); grid on;
end


return

%% Create the output for JSON config
waveName = 'Schroeder';
WaveDef = [];
for iChan = 1:structMultiSine.numChan
    name = [waveName '_' num2str(iChan)];
    WaveDef.(name).Type = 'MultiSine';
    WaveDef.(name).Duration = structMultiSine.timeDur_s;
    WaveDef.(name).Frequency = structMultiSine.freqChan_rps{iChan};
    WaveDef.(name).Phase = structMultiSine.phaseChan_rad{iChan};
    WaveDef.(name).Amplitude = structMultiSine.ampChan_nd{iChan};
end

jsonStr = jsonencode(WaveDef);

% Pretty Print
jsonStr = strrep(jsonStr, '},"', sprintf('},\r"'));
jsonStr = strrep(jsonStr, ':{', sprintf(':{\r\t'));
jsonStr = strrep(jsonStr, ',"', sprintf(',\r\t"'));
jsonStr = strrep(jsonStr, '}}', sprintf('}\r}'));
jsonStr = strrep(jsonStr, ']}', sprintf(']\r}'));
jsonStr = strrep(jsonStr, ':', sprintf(': '));
jsonStr = strrep(jsonStr, ',', sprintf(', '));

disp(jsonStr)
